From e5b9a205d65053d56b9d32bdf0f2806972fb472c Mon Sep 17 00:00:00 2001
From: David Rubin <david@vortan.dev>
Date: Thu, 6 Nov 2025 20:27:38 -0800
Subject: [PATCH 047/133] llvm: add extra clobbers to valgrind requests
X-Developer-Signature: v=1; a=openpgp-sha256; l=2160; i=xtex@astrafall.org;
 h=from:subject; bh=/jlbTKR87kmgSGvPLuGLh0bpPR7c/oySeCA5Wu8QU+0=;
 b=owGbwMvMwCW2U4Ij7wZL9ETG02pJDJn78hhadp7ctpTh1ZxpYXfanIU5A870ZCm4ionYZgh8P
 eD6syW4YyILgxgXg6WYIkuRYYM3q046v+iyclmYOaxMIEOkRRoYGBgYWBj4chPzSo10jPRMtQ31
 DA11jHSMGLg4BWCqV+gy/BVfInnooEVF/jaOWR4LFyvwh7W3zZRt6TyXsHe2+m7PzE+MDJt3/El
 asiWm+rCl918OFle/+R1X/6ZP2Thrw6HF6UZvorgB
X-Developer-Key: i=xtex@astrafall.org; a=openpgp;
 fpr=7231804B052C670F15A6771DB918086ED8045B91

This seems to work around a very puzzling miscompilation first
present in LLVM 21.x. We already unconditionally add these
clobbers to inline assembly that came from the source, the
valgrind requests should also contain them.

(cherry picked from commit 483f9bd3670a52242d799dd4eee4273bdc714e8a)
---
 src/codegen/llvm.zig | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/src/codegen/llvm.zig b/src/codegen/llvm.zig
index 1a9d1d869b2d..b2f1518caa07 100644
--- a/src/codegen/llvm.zig
+++ b/src/codegen/llvm.zig
@@ -11742,7 +11742,7 @@ pub const FuncGen = struct {
                 \\ srl $$0,  $$0,  19
                 \\ or  $$13, $$13, $$13
                 ,
-                .constraints = "={$11},{$12},{$11},~{memory}",
+                .constraints = "={$11},{$12},{$11},~{memory},~{$1}",
             },
             .mips64, .mips64el => .{
                 .template =
@@ -11750,7 +11750,7 @@ pub const FuncGen = struct {
                 \\ dsll $$0,  $$0,  29   ; dsll $$0, $$0, 19
                 \\ or   $$13, $$13, $$13
                 ,
-                .constraints = "={$11},{$12},{$11},~{memory}",
+                .constraints = "={$11},{$12},{$11},~{memory},~{$1}",
             },
             .powerpc, .powerpcle => .{
                 .template =
@@ -11797,7 +11797,7 @@ pub const FuncGen = struct {
                 \\ roll  $$61, %edi ; roll $$51, %edi
                 \\ xchgl %ebx, %ebx
                 ,
-                .constraints = "={edx},{eax},{edx},~{cc},~{memory}",
+                .constraints = "={edx},{eax},{edx},~{cc},~{memory},~{dirflag},~{fpsr},~{flags}",
             },
             .x86_64 => .{
                 .template =
@@ -11805,7 +11805,7 @@ pub const FuncGen = struct {
                 \\ rolq  $$61, %rdi ; rolq $$51, %rdi
                 \\ xchgq %rbx, %rbx
                 ,
-                .constraints = "={rdx},{rax},{rdx},~{cc},~{memory}",
+                .constraints = "={rdx},{rax},{rdx},~{cc},~{memory},~{dirflag},~{fpsr},~{flags}",
             },
             else => unreachable,
         };
-- 
2.52.0

